src/java.base/share/native/libjli/java.c

```diff
@@ -525,6 +529,17 @@ JavaMain(void* _args)
     appClass = GetApplicationClass(env);
     NULL_CHECK_RETURN_VALUE(appClass, -1);
 
+    /*  Get launcher helper class. */
+    helperClass = GetLauncherHelperClass(env);
+
+    /*  Validate and fetch main method from application class. */
+    validateMainMethod = (*env)->GetStaticMethodID(env, helperClass,
+                                         "validateMainMethod",
+                                         "(Ljava/lang/Class;)Ljava/lang/reflect/Method;");
+    CHECK_EXCEPTION_NULL_LEAVE(validateMainMethod);
+    mainMethod = (*env)->CallStaticObjectMethod(env, helperClass, validateMainMethod, mainClass);
+    NULL_CHECK_RETURN_VALUE(mainMethod, -1);
+
     /* Build platform specific argument array */
     mainArgs = CreateApplicationArgs(env, argv, argc);
```

Comment:
we build the arguments always even if the main we will end up invoking receives no argument and thus these arguments will be discarded. Could it make sense to create the arguments only if they won't be discarded?

Reply:
Sure.